controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,361 @@
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/builder"
+
+	"testing"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestClusterReconciler_reconcile(t *testing.T) {
+	g := NewWithT(t)
+	timeout := 5 * time.Second
+	ns, err := env.CreateNamespace(ctx, "test-machine-watches")
+	g.Expect(err).ToNot(HaveOccurred())
+	clusterName := "cluster1"
+	clusterClassName := "class1"
+	workerClassName := "linux-worker"
+	infrastructureMachineTemplate := builder.InfrastructureMachineTemplate(ns.Name, "inframachinetemplate").Build()
+	infrastructureCluster := builder.InfrastructureClusterTemplate(ns.Name, "infraclustertemplate").
+		// Create spec fake setting to ensure that template spec is non-empty for tests.
+		WithSpecFields(map[string]interface{}{"spec.template.spec.fakeSetting": true}).
+		Build()
+	controlPlane := builder.ControlPlaneTemplate(ns.Name, "cp1").
+		WithInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		Build()
+	bootstrapTemplate := builder.BootstrapTemplate(ns.Name, "bootstraptemplate").Build()
+	machineDeploymentTopology := builder.MachineDeploymentTopology(workerClassName).
+		WithName("mdm1").
+		WithReplicas(3).
+		Build()
+	machineDeploymentTopology2 := builder.MachineDeploymentTopology(workerClassName).
+		WithName("mdm2").
+		WithReplicas(1).
+		Build()
+	machineDeploymentClass := builder.MachineDeploymentClass(ns.Name, "md1").
+		WithClass(workerClassName).
+		WithInfrastructureTemplate(infrastructureMachineTemplate).
+		WithBootstrapTemplate(bootstrapTemplate).
+		WithLabels(map[string]string{"foo": "bar"}).
+		WithAnnotations(map[string]string{"foo": "bar"}).
+		Build()
+	clusterClass := builder.ClusterClass(ns.Name, clusterClassName).
+		WithInfrastructureClusterTemplate(infrastructureCluster).
+		WithControlPlaneTemplate(controlPlane).
+		WithControlPlaneInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass}).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology(clusterClass.Name).
+				WithWorkerTopology(machineDeploymentTopology).
+				WithWorkerTopology(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects created above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureCluster,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlane,
+	}
+
+	testName := "Cluster creation from barebones cluster object and clusterClass."
+	t.Run(testName, func(t *testing.T) {
+		defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+		for _, obj := range initObjs {
+			err := env.Create(ctx, obj)
+			g.Expect(err).ToNot(HaveOccurred())
```

Comment:
should we have a defer clause or sync func that ensures and validate all these CRs are deleted succesfully?